README
$Id: README 115 2008-12-01 21:17:51Z swistow $

This is XML::Feed, an abstraction above the RSS and Atom syndication
feed formats. It supports both parsing and autodiscovery of feeds.

PREREQUISITES

    * Class::ErrorHandler
    * XML::RSS
    * XML::Atom
    * DateTime
    * DateTime::Format::Mail
    * DateTime::Format::W3CDTF
    * List::Util
    * Feed::Find
    * URI::Fetch

INSTALLATION

XML::Feed installation is straightforward. If your CPAN shell
is set up, you should just be able to do

    % perl -MCPAN -e 'install XML::Feed'

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

Six Apart / cpan@sixapart.com


Changes
2013-02-28  Dave Cross <dave@dave.org.uk>

  * lib/XML/Feed.pm, lib/XML/Feed/Content.pm, lib/XML/Feed/Enclosure.pm,
  lib/XML/Feed/Entry.pm, lib/XML/Feed/Entry/Format/Atom.pm,
  lib/XML/Feed/Entry/Format/RSS.pm, lib/XML/Feed/Format/Atom.pm,
  lib/XML/Feed/Format/RSS.pm: Bump versions for release.

2013-02-26  Gabor Szabo <szabgab@gmail.com>

  * Build.PL: add dist_author and x_contributors to META files

  * .gitignore: ignore some generated files

2013-01-08  Gabor Szabo <szabgab@gmail.com>

  * lib/XML/Feed.pm, lib/XML/Feed/Content.pm, lib/XML/Feed/Enclosure.pm,
  lib/XML/Feed/Entry.pm, lib/XML/Feed/Entry/Format/Atom.pm,
  lib/XML/Feed/Entry/Format/RSS.pm, lib/XML/Feed/Format/Atom.pm,
  lib/XML/Feed/Format/RSS.pm: add use warnings to the modules

  * Build.PL, lib/XML/Feed/Content.pm, lib/XML/Feed/Entry.pm,
  lib/XML/Feed/Format/Atom.pm, lib/XML/Feed/Format/RSS.pm, t/00-compile.t,
  t/01-parse.t, t/04-splice-atom.t, t/04-splice-rss.t,
  t/08-convert-summary-bug.t, t/10-mix-and-match.t, t/11-xml-base-atom.t,
  t/11-xml-base-rss.t, t/16-convert-mult-categories.t: remove Id tags from the
  times of Subversion

  * lib/XML/Feed/Content.pm, lib/XML/Feed/Enclosure.pm, lib/XML/Feed/Entry.pm,
  lib/XML/Feed/Entry/Format/Atom.pm, lib/XML/Feed/Entry/Format/RSS.pm,
  lib/XML/Feed/Format/Atom.pm, lib/XML/Feed/Format/RSS.pm, t/has-version.t: add
  version number to all the modules

2013-01-04  Dave Cross <dave@dave.org.uk>

  * lib/XML/Feed.pm: Bump version for release

  * Changes: Commit changes to Changes

2012-12-30  Gabor Szabo <szabgab@gmail.com>

  * Changes, lib/XML/Feed.pm: change the User Agent string

2012-07-28  Shlomi Fish <shlomif@shlomifish.org>

  * Build.PL: Add a meta_merge link to the respository.  This way metacpan.org
  and other services will display it.

2012-07-10  Dave Cross <dave@dave.org.uk>

  * lib/XML/Feed.pm: Bump version for release.

  * lib/XML/Feed.pm, lib/XML/Feed/Format/Atom.pm, lib/XML/Feed/Format/RSS.pm:
  Added an image() method to Feed objects. Does something useful for RSS feeds;
  not for Atom feeds currently.

  * t/02-create.t: Fixed tests

2012-04-10  Jakob Voss <voss@gbv.de>

  * Changes, lib/XML/Feed.pm, lib/XML/Feed/Format/Atom.pm,
  lib/XML/Feed/Format/RSS.pm, t/02-create.t: Add support for RFC 5005 Feed
  Paging and Archiving in Atom feeds

2012-04-06  Dave Cross <dave@dave.org.uk>

  * Changes, lib/XML/Feed.pm: Bump version for release.

2012-04-05  Dave Cross <dave@dave.org.uk>

  * lib/XML/Feed.pm: Remove svn tag.

  * lib/XML/Feed.pm: Changed contact details from non-working 6A email address.

2012-03-28  sockmonk <wes.sheldahl@gmail.com>

  * README: Updated README file with current and tested installation
  instructions.  Hoping to clarify the confusion reflected in RT # 53884, maybe
  even resolve it.  https://rt.cpan.org/Public/Bug/Display.html?id=53884

2012-03-12  Dave Cross <dave@dave.org.uk>

  * lib/XML/Feed.pm: Bump version for release.

  * t/07-atom10-create.t: Fixed broken date test.

  * lib/XML/Feed/Entry/Format/Atom.pm: Applying the missing bit of the patch
  from https://rt.cpan.org/Ticket/Display.html?id=20763

2012-03-10  Dave Cross <dave@dave.org.uk>

  * lib/XML/Feed.pm: Bump version number for release.

  * Build.PL: Increase XML::Atom dependency to 0.37 - see
  https://rt.cpan.org/Ticket/Display.html?id=74703

2012-03-06  Gabor Szabo <szabgab@gmail.com>

  * MANIFEST.SKIP: fix MANIFEST.SKIP

  * MANIFEST: update MANIFEST file

  * eg/check_feed.pl: add example

  * lib/XML/Feed.pm, lib/XML/Feed/Entry.pm: add links to the POD

2012-02-19  Gabor Szabo <szabgab@gmail.com>

  * t/23-eval.t: add explicit tests for loading internal packages and failing
  when it does not exist

  * lib/XML/Feed/Entry/Format/Atom.pm, lib/XML/Feed/Entry/Format/RSS.pm,
  lib/XML/Feed/Format/Atom.pm, lib/XML/Feed/Format/RSS.pm: split up files to
  individual packages

  * .gitignore: ignore *.swp files

2012-02-18  Gabor Szabo <szabgab@gmail.com>

  * .gitignore: gitignore some generated files

2012-02-17  Gabor Szabo <szabgab@gmail.com>

  * lib/XML/Feed/Format/RSS.pm: add missing return call fixing the failures
  introduced when trying to fix 73160

  * lib/XML/Feed/Format/RSS.pm: eliminate warning generated in
  t/16-convert-mult-categories.t

2011-12-11  Dave Cross <dave@dave.org.uk>

  * MANIFEST, lib/XML/Feed/Format/RSS.pm, t/22-bug73160.t,
  t/samples/rss10-datespaces.xml: Added tests for bug 73160. And code to make
  them pass.

2011-09-03  Dave Cross <dave@dave.org.uk>

  * Changes: Updated changed for 0.45 (oops) and 0.46.

  * MANIFEST, lib/XML/Feed.pm, lib/XML/Feed/Format/Atom.pm,
  lib/XML/Feed/Format/RSS.pm, t/20-no-enclosures.t, t/21-rss2-permalinks.t,
  t/samples/rss20-p.xml: Closed a few tickets from the RT queue as an excuse
  for a new release. Bumped to version 0.46.

2011-07-30  Dave Cross <dave@dave.org.uk>

  * MANIFEST, MANIFEST.SKIP, lib/XML/Feed.pm, lib/XML/Feed/Format/Atom.pm,
  t/18-double2.t, t/19-double3.t, t/samples/rss20-double.xml: Another round of
  changes to fix (and test)
  https://rt.cpan.org/Public/Bug/Display.html?id=44899

2011-07-16  Dave Cross <dave@dave.org.uk>

  * MANIFEST, lib/XML/Feed/Format/Atom.pm, t/19-double3.t,
  t/samples/rss10-double2.xml: Added new (currently failing) test for the
  double-encoding bug.

2011-04-24  Dave Cross <dave@dave.org.uk>

  * MANIFEST, MANIFEST.SKIP, t/18-double2.t, t/samples/rss20-double.xml: Tests
  to ensure that the double-encoding bug fix works.

2011-03-13  Dave Cross <dave@dave.org.uk>

  * t/17-double.t, t/18-double2.t, t/samples/rss10-double.xml,
  t/samples/rss20-double.xml: Added new tests to prove what the problem is.

  * t/15-odd-date.t: Force stringification so test passes.

2010-03-22  Simon Wistow <simon@thegestalt.org>

  * Build.PL: Force requiring HTML::Entities;

  * Build.PL, Changes, lib/XML/Feed.pm, lib/XML/Feed/Format/Atom.pm: Fix bug in
  conversion process between RSS and Atom where HTML got double encoded 

Makefile.PL
# Note: this file was auto-generated by Module::Build::Compat version 0.4003
use ExtUtils::MakeMaker;
WriteMakefile
(
  'NAME' => 'XML::Feed',
  'VERSION_FROM' => 'lib/XML/Feed.pm',
  'PREREQ_PM' => {
                   'Class::ErrorHandler' => 0,
                   'DateTime' => 0,
                   'DateTime::Format::Mail' => 0,
                   'DateTime::Format::W3CDTF' => 0,
                   'Feed::Find' => 0,
                   'HTML::Entities' => 0,
                   'HTML::TokeParser' => 0,
                   'LWP::UserAgent' => 0,
                   'List::Util' => 0,
                   'Module::Pluggable' => 0,
                   'Test::More' => 0,
                   'URI::Fetch' => 0,
                   'XML::Atom' => '0.38',
                   'XML::LibXML' => '1.66',
                   'XML::RSS' => '1.47'
                 },
  'INSTALLDIRS' => 'site',
  'EXE_FILES' => [],
  'PL_FILES' => {}
)
;


lib/XML/Feed.pm
# $Id: Feed.pm 150 2009-05-07 00:47:58Z swistow $

package XML::Feed;
use strict;

use base qw( Class::ErrorHandler );
use Feed::Find;
use URI::Fetch;
use LWP::UserAgent;
use Carp;
use Module::Pluggable search_path => "XML::Feed::Format",
                      require     => 1,
                      sub_name    => 'formatters';

our $VERSION = '0.43';
our $MULTIPLE_ENCLOSURES = 0;
our @formatters;
BEGIN {
	@formatters = __PACKAGE__->formatters;
}

sub new {
    my $class = shift;
    my $format = shift || 'Atom';
    my $format_class = 'XML::Feed::Format::' . $format;
    eval "use $format_class";
    Carp::croak("Unsupported format $format: $@") if $@;
    my $feed = bless {}, join('::', __PACKAGE__, "Format", $format);
    $feed->init_empty(@_) or return $class->error($feed->errstr);
    $feed;
}

sub init_empty { 1 }

sub parse {
    my $class = shift;
    my($stream, $specified_format) = @_;
    return $class->error("Stream parameter is required") unless $stream;
    my $feed = bless {}, $class;
    my $xml = '';
    if (UNIVERSAL::isa($stream, 'URI')) {
        my $ua  = LWP::UserAgent->new;
        $ua->env_proxy; # force allowing of proxies
        my $res = URI::Fetch->fetch($stream, UserAgent => $ua)
            or return $class->error(URI::Fetch->errstr);
        return $class->error("This feed has been permanently removed")
            if $res->status == URI::Fetch::URI_GONE();
        $xml = $res->content;
    } elsif (ref($stream) eq 'SCALAR') {
        $xml = $$stream;
    } elsif (ref($stream)) {
        while (read($stream, my($chunk), 8192)) {
            $xml .= $chunk;
        }
    } else {
        open my $fh, $stream
            or return $class->error("Can't open $stream: $!");
        while (read $fh, my($chunk), 8192) {
            $xml .= $chunk;
        }
        close $fh;
    }
    return $class->error("Can't get feed XML content from $stream")
        unless $xml;
    my $format;
    if ($specified_format) {
        $format = $specified_format;
    } else {
        $format = $feed->identify_format(\$xml) or return $class->error($feed->errstr);
    }

    my $format_class = join '::', __PACKAGE__, "Format", $format;
    eval "use $format_class";
    return $class->error("Unsupported format $format: $@") if $@;
    bless $feed, $format_class;
    $feed->init_string(\$xml) or return $class->error($feed->errstr);
    $feed;
}

sub identify_format {
    my $feed   = shift;
    my($xml)   = @_;
	foreach my $class (@formatters) {
		my ($name) = ($class =~ m!([^:]+)$!);
		# TODO ugly
		my $tmp = $$xml;
		return $name if eval { $class->identify(\$tmp) };
		return $feed->error($@) if $@;
	} 
	return $feed->error("Cannot detect feed type");
}

sub _get_first_tag {
	my $class  = shift;
	my ($xml)  = @_;


    ## Auto-detect feed type based on first element. This is prone
    ## to breakage, but then again we don't want to parse the whole
    ## feed ourselves.
    my $tag;
    while ($$xml =~ /<(\S+)/sg) {
        (my $t = $1) =~ tr/a-zA-Z0-9:\-\?!//cd;
        my $first = substr $t, 0, 1;
        $tag = $t, last unless $first eq '?' || $first eq '!';
    }
	die ("Cannot find first element") unless $tag;
    $tag =~ s/^.*://;
	return $tag;
}

sub find_feeds {
    my $class = shift;
    my($uri) = @_;
    my @feeds = Feed::Find->find($uri)
        or return $class->error(Feed::Find->errstr);
    @feeds;
}

sub convert {
    my $feed = shift;
    my($format) = @_;
    my $new = XML::Feed->new($format);
    for my $field (qw( title link description language author copyright modified generator )) {
        my $val = $feed->$field();
        next unless defined $val;
        $new->$field($val);
    }
    for my $entry ($feed->entries) {
        $new->add_entry($entry->convert($format));
    }
    $new;
}

sub splice {
    my $feed = shift;
    my($other) = @_;
    my %ids = map { $_->id => 1 } $feed->entries;
    for my $entry ($other->entries) {
        $feed->add_entry($entry) unless $ids{$entry->id}++;
    }
}

sub _convert_entry {
    my $feed   = shift;
    my $entry  = shift;
    my $feed_format  = ref($feed);   $feed_format  =~ s!^XML::Feed::Format::!!;
    my $entry_format = ref($entry);  $entry_format =~ s!^XML::Feed::Entry::Format::!!;
    return $entry if $entry_format eq $feed_format;
    return $entry->convert($feed_format); 
}

sub base;
sub format;
sub title;
sub link;
sub self_link;
sub description;
sub language;
sub author;
sub copyright;
sub modified;
sub generator;
sub add_entry;
sub entries;
sub as_xml;
sub id;

sub tagline { shift->description(@_) }
sub items   { $_[0]->entries     }

1;
__END__

=head1 NAME

XML::Feed - Syndication feed parser and auto-discovery

=head1 SYNOPSIS

    use XML::Feed;
    my $feed = XML::Feed->parse(URI->new('http://example.com/atom.xml'))
        or die XML::Feed->errstr;
    print $feed->title, "\n";
    for my $entry ($feed->entries) {
    }

    ## Find all of the syndication feeds on a given page, using
    ## auto-discovery.
    my @feeds = XML::Feed->find_feeds('http://example.com/');

=head1 DESCRIPTION

I<XML::Feed> is a syndication feed parser for both RSS and Atom feeds. It
also implements feed auto-discovery for finding feeds, given a URI.

I<XML::Feed> supports the following syndication feed formats:

=over 4

=item * RSS 0.91

=item * RSS 1.0

=item * RSS 2.0

=item * Atom

=back

The goal of I<XML::Feed> is to provide a unified API for parsing and using
the various syndication formats. The different flavors of RSS and Atom
handle data in different ways: date handling; summaries and content;
escaping and quoting; etc. This module attempts to remove those differences
by providing a wrapper around the formats and the classes implementing
those formats (I<XML::RSS> and I<XML::Atom::Feed>). For example, dates are
handled differently in each of the above formats. To provide a unified API for
date handling, I<XML::Feed> converts all date formats transparently into
I<DateTime> objects, which it then returns to the caller.

=head1 USAGE

=head2 XML::Feed->new($format)

Creates a new empty I<XML::Feed> object using the format I<$format>.

    $feed = XML::Feed->new('Atom');
    $feed = XML::Feed->new('RSS');
    $feed = XML::Feed->new('RSS', version => '0.91');

=head2 XML::Feed->parse($stream)

=head2 XML::Feed->parse($stream, $format)

Parses a syndication feed identified by I<$stream>. I<$stream> can be any
one of the following:

=over 4

=item * Scalar reference

A reference to string containing the XML body of the feed.

=item * Filehandle

An open filehandle from which the feed XML will be read.

=item * File name

The name of a file containing the feed XML.

=item * URI object

A URI from which the feed XML will be retrieved.

=back

I<$format> allows you to override format guessing.

=head2 XML::Feed->find_feeds($uri)

Given a URI I<$uri>, use auto-discovery to find all of the feeds linked
from that page (using I<E<lt>linkE<gt>> tags).

Returns a list of feed URIs.

=head2 XML::Feed->identify_format($xml)

Given the xml of a feed return what format it is in (C<Atom>, or some version of C<RSS>).

=head2 $feed->convert($format)

Converts the I<XML::Feed> object into the I<$format> format, and returns
the new object.

=head2 $feed->splice($other_feed)

Splices in all of the entries from the feed I<$other_feed> into I<$feed>,
skipping posts that are already in I<$feed>.

=head2 $feed->format

Returns the format of the feed (C<Atom>, or some version of C<RSS>).

=head2 $feed->title([ $title ])

The title of the feed/channel.

=head2 $feed->base([ $base ])

The url base of the feed/channel.

=head2 $feed->link([ $uri ])

The permalink of the feed/channel.

=head2 $feed->tagline([ $tagline ])

The description or tagline of the feed/channel.

=head2 $feed->description([ $description ])

Alias for I<$feed-E<gt>tagline>.

=head2 $feed->author([ $author ])

The author of the feed/channel.

=head2 $feed->language([ $language ])

The language of the feed.

=head2 $feed->copyright([ $copyright ])

The copyright notice of the feed.

=head2 $feed->modified([ $modified ])

A I<DateTime> object representing the last-modified date of the feed.

If present, I<$modified> should be a I<DateTime> object.

=head2 $feed->generator([ $generator ])

The generator of the feed.

=head2 $feed->self_link ([ $uri ])

The Atom Self-link of the feed:

L<http://validator.w3.org/feed/docs/warning/MissingAtomSelfLink.html>

A string.

=head2 $feed->entries

A list of the entries/items in the feed. Returns an array containing
I<XML::Feed::Entry> objects.

=head2 $feed->items

A synonym for I<$feed->entries>.

=head2 $feed->add_entry($entry)

Adds an entry to the feed. I<$entry> should be an I<XML::Feed::Entry>
object in the correct format for the feed.

=head2 $feed->as_xml

Returns an XML representation of the feed, in the format determined by
the current format of the I<$feed> object.

=head1 PACKAGE VARIABLES

=over 4

=item C<$XML::Feed::RSS::PREFERRED_PARSER>

If you want to use another RSS parser class than XML::RSS (default), you can
change the class by setting C<$PREFERRED_PARSER> variable in XML::Feed::RSS
package.

    $XML::Feed::RSS::PREFERRED_PARSER = "XML::RSS::LibXML";

B<Note:> this will only work for parsing feeds, not creating feeds.

B<Note:> Only C<XML::RSS::LibXML> version 0.3004 is known to work at the moment.

=item C<$XML::Feed::MULTIPLE_ENCLOSURES>

Although the RSS specification states that there can be at most one enclosure per item 
some feeds break this rule.

If this variable is set then C<XML::Feed> captures all of them and makes them available as a list.

Otherwise it returns the last enclosure parsed.

B<Note:> C<XML::RSS> version 1.44 is needed for this to work.

=back

=cut

=head1 VALID FEEDS

For reference, this cgi script will create valid, albeit nonsensical feeds 
(according to C<http://feedvalidator.org> anyway) for Atom 1.0 and RSS 0.90, 
0.91, 1.0 and 2.0. 

    #!perl -w

    use strict;
    use CGI;
    use CGI::Carp qw(fatalsToBrowser);
    use DateTime;
    use XML::Feed;

    my $cgi  = CGI->new;
    my @args = ( $cgi->param('format') || "Atom" );
    push @args, ( version => $cgi->param('version') ) if $cgi->param('version');

    my $feed = XML::Feed->new(@args);
    $feed->id("http://".time.rand()."/");
    $feed->title('Test Feed');
    $feed->link($cgi->url);
    $feed->self_link($cgi->url( -query => 1, -full => 1, -rewrite => 1) );
    $feed->modified(DateTime->now);

    my $entry = XML::Feed::Entry->new();
    $entry->id("http://".time.rand()."/");
    $entry->link("http://example.com");
    $entry->title("Test entry");
    $entry->summary("Test summary");
    $entry->content("Foo");
    $entry->modified(DateTime->now);
    $entry->author('test@example.com (Testy McTesterson)');
    $feed->add_entry($entry);

    my $mime = ("Atom" eq $feed->format) ? "application/atom+xml" : "application/rss+xml";
    print $cgi->header($mime);
    print $feed->as_xml;


=head1 LICENSE

I<XML::Feed> is free software; you may redistribute it and/or modify it
under the same terms as Perl itself.

=head1 AUTHOR & COPYRIGHT

Except where otherwise noted, I<XML::Feed> is Copyright 2004-2008
Six Apart, cpan@sixapart.com. All rights reserved.

=head1 SUBVERSION 

The latest version of I<XML::Feed> can be found at

    http://code.sixapart.com/svn/XML-Feed/trunk/

=cut


lib/XML/Feed/Entry.pm
# $Id: Entry.pm 118 2008-12-10 20:15:44Z swistow $

package XML::Feed::Entry;
use strict;
use base qw( Class::ErrorHandler );

use Scalar::Util qw( blessed );

use Carp;

sub wrap {
    my $class = shift;
    my($item) = @_;
    bless { entry => $item }, $class;
}

sub unwrap { $_[0]->{entry} }

sub new {
    my $class = shift;
    my($format) = @_;
    $format ||= 'Atom';
    my $format_class = 'XML::Feed::Format::' . $format;
    eval "use $format_class";
    Carp::croak("Unsupported format $format: $@") if $@;
    my $entry = bless {}, join('::', __PACKAGE__, "Format", $format);
    $entry->init_empty or return $class->error($entry->errstr);
    $entry;
}

sub init_empty { 1 }

sub convert {
    my $entry = shift;
    my($format) = @_;
    my $new = __PACKAGE__->new($format);
    for my $field (qw( title link content summary category author id issued modified lat long )) {
        my $val = $entry->$field();
        next unless defined $val;
        next if blessed $val && $val->isa('XML::Feed::Content') && ! defined $val->body;
        $new->$field($val);
    }
    $new;
}

sub title;
sub link;
sub content;
sub summary;
sub category;
sub author;
sub id;
sub issued;
sub modified;
sub lat;
sub long;
sub format;
sub tags { shift->category(@_) }

1;
__END__

=head1 NAME

XML::Feed::Entry - Entry/item in a syndication feed

=head1 SYNOPSIS

    ## $feed is an XML::Feed object.
    for my $entry ($feed->entries) {
        print $entry->title, "\n", $entry->summary, "\n\n";
    }

=head1 DESCRIPTION

I<XML::Feed::Entry> represents an entry/item in an I<XML::Feed> syndication
feed.

=head1 USAGE

=head2 XML::Feed::Entry->new($format)

Creates a new I<XML::Feed::Entry> object in the format I<$format>, which
should be either I<RSS> or I<Atom>.

=head2 $entry->convert($format)

Converts the I<XML::Feed::Entry> object into the I<$format> format, and
returns the new object.

=head2 $entry->title([ $title ])

The title of the entry.

=head2 $entry->base([ $base ])

The url base of the entry.

=head2 $entry->link([ $uri ])

The permalink of the entry, in most cases, except in cases where it points
instead to an offsite URI referenced in the entry.

=head2 $entry->content([ $content ])

Bn I<XML::Feed::Content> object representing the full entry body, or as
much as is available in the feed.

In RSS feeds, this method will look first for
I<http://purl.org/rss/1.0/modules/content/#encoded> and
I<http://www.w3.org/1999/xhtml#body> elements, then fall back to a
I<E<lt>descriptionE<gt>> element.

=head2 $entry->summary([ $summary ])

An I<XML::Feed::Content> object representing a short summary of the entry.
Possibly.

Since RSS feeds do not have the idea of a summary separate from the entry
body, this may not always be what you want. If the entry contains both a
I<E<lt>descriptionE<gt>> element B<and> another element typically used for
the full content of the entry--either I<http://www.w3.org/1999/xhtml/body>
or I<http://purl.org/rss/1.0/modules/content/#encoded>--we treat that as
the summary. Otherwise, we assume that there isn't a summary, and return
an I<XML::Feed::Content> object with an empty string in the I<body>.

=head2 $entry->category([ $category ])

The category in which the entry was posted.

Returns a list of categories if called in array context or the first
category if called in scalar context.

B<WARNING> It's possible this API might change to have an 
I<add_category> instead.

=head2 $entry->tags([ $tag ])

A synonym for I<category>;

=head2 $entry->author([ $author ])

The name or email address of the person who posted the entry.

=head2 $entry->id([ $id ])

The unique ID of the entry.

=head2 $entry->issued([ $issued ])

A I<DateTime> object representing the date and time at which the entry
was posted.

If present, I<$issued> should be a I<DateTime> object.

=head2 $entry->modified([ $modified ])

A I<DateTime> object representing the last-modified date of the entry.

If present, I<$modified> should be a I<DateTime> object.

=head2 $entry->wrap

Take an entry in its native format and turn it into an I<XML::Feed::Entry> object.

=head2 $entry->unwrap

Take an I<XML::Feed::Entry> object and turn it into its native format.

=head1 AUTHOR & COPYRIGHT

Please see the I<XML::Feed> manpage for author, copyright, and license
information.

=cut


Build.PL
# $Id: Build.PL 110 2008-11-24 00:43:50Z swistow $
use Module::Build;

my $build = Module::Build
  ->new( module_name   => "XML::Feed",
         dist_abstract => "XML Syndication Feed Support", 
         license       => 'perl',
         script_files  => [ ],
         requires      => {
                          'Class::ErrorHandler'        => 0,
                          'DateTime'                   => 0,
                          'DateTime::Format::Mail'     => 0,
                          'DateTime::Format::W3CDTF'   => 0,
                          'Feed::Find'                 => 0,
                          'HTML::TokeParser'           => 0,
                          'List::Util'                 => 0,
                          'LWP::UserAgent'             => 0,
                          'Module::Pluggable'          => 0,
                          'URI::Fetch'                 => 0,
                          'XML::Atom'                  => 0.32,
                          'XML::LibXML'                => 1.66,
                          'XML::RSS'                   => 1.37,
                          'Test::More'                 => 0,
                        },
         create_makefile_pl => 'traditional',
       );

$build->create_build_script;


